### ISSUE TITLE
[CVE-2015-20107] mailcap.findmatch: document shell command Injection danger in filename parameter

### ISSUE THREAD

**Issue by 3077c527-db90-456b-9bc4-abba055100b5** [python/cpython#68966]
BPO | [24778](https://bugs.python.org/issue24778)
--- | :---
Nosy | @vstinner, @bitdancer
Files | <li>[screenshot.png](https://bugs.python.org/file40099/screenshot.png "Uploaded as image/png at 2015-08-02.08:25:06 by TheRegRunner")</li><li>[The Quote Problem.py](https://bugs.python.org/file40116/The%20Quote%20Problem.py "Uploaded as text/plain at 2015-08-03.19:27:35 by TheRegRunner")</li><li>[mailcap patch.zip](https://bugs.python.org/file40897/mailcap%20patch.zip "Uploaded as application/zip at 2015-10-29.19:26:36 by TheRegRunner"): mailcap.py patches and diffs for python2.7 and python 3.5</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2015-08-02.08:25:07.171>
labels = ['type-security', '3.11', 'library', 'docs']
title = 'mailcap.findmatch: document shell command Injection danger in filename parameter'
updated_at = <Date 2022-04-06.15:30:37.106>
user = 'https://bugs.python.org/TheRegRunner'
```

bugs.python.org fields:
```python
activity = <Date 2022-04-06.15:30:37.106>
actor = 'vstinner'
assignee = 'docs@python'
closed = False
closed_date = None
closer = None
components = ['Documentation', 'Library (Lib)']
creation = <Date 2015-08-02.08:25:07.171>
creator = 'TheRegRunner'
dependencies = []
files = ['40099', '40116', '40897']
hgrepos = []
issue_num = 24778
keywords = []
message_count = 14.0
messages = ['247857', '247861', '247944', '247946', '247951', '247979', '247992', '248058', '248061', '248062', '248070', '248074', '253689', '416878']
nosy_count = 4.0
nosy_names = ['vstinner', 'r.david.murray', 'docs@python', 'TheRegRunner']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'security'
url = 'https://bugs.python.org/issue24778'
versions = ['Python 3.11']
```

</p></details>

**Comment by zooba** [comment]
It's too bad this didn't make it into [PEP 594](https://peps.python.org/pep-0594/), but I think we can deprecate it anyway unless someone steps up to be a maintainer.

The original patch doesn't add quotes in a way that solves the problem without potentially breaking legitimate users. The best mitigation is to validate user-provided values before using them, in this case, probably by checking that the file exists.

**Comment by zooba** [comment]
FYI, this issue now has the CVE ID `CVE-2015-20107` to assist with discussions.

Notification at https://mail.python.org/archives/list/security-announce@python.org/thread/QDSXNCW77UGULFG2JMDFZQ7H4DIR32LA/

**Comment by brettcannon** [comment]
I vote to deprecate. I guess we just need to ask on python-dev/committers to see if anyone will maintain it and what the community impact may be?

**Comment by brettcannon** [comment]
I've proposed deprecating mailcap at https://mail.python.org/archives/list/python-dev@python.org/thread/EB2BS4DBWSTBIOPQL5QTBSIOBORWSCMJ/ .

**Comment by arhadthedev** [comment]
@brettcannon Indeed, 3.10-3.12 need to get a fix.

![release phases on a timeline for Python 3.9-3.12](https://peps.python.org/_images/pep-0602-example-release-calendar.png)
*<sup>Figure 1 from [PEP 602 Annual Release Cycle for Python](https://peps.python.org/pep-0602/)</sup>*

**Comment by gpshead** [comment]
@CySirX Please do not post screenshots containing textual communications. Just the text or the link to the document. Bitmaps are unsearchable and unaccessible and always render wrong compared to the readers actual display.

**Comment by vstinner** [comment]
This issue can be worked around in the caller. For example, by creating a temporary file with a safe filename.

**Comment by gpshead** [comment]
FWIW the only third party package we've found using mailcap already does create a safe filename before calling mailcap: https://github.com/mitmproxy/mitmproxy/blob/main/mitmproxy/tools/console/master.py#L153

**Comment by hugovk** [comment]
> It's too bad this didn't make it into [PEP 594](https://peps.python.org/pep-0594/), but I think we can deprecate it anyway unless someone steps up to be a maintainer.

Update: mailcap has now been added to PEP 594 and will be deprecated in Python 3.11 and removed in 3.13:

* [PEP 594 – Removing dead batteries from the standard library](https://peps.python.org/pep-0594/)
* https://github.com/python/cpython/issues/91217

**Comment by vstinner** [comment]
I created https://github.com/python/cpython/pull/91951 to deprecate mailcap.

**Comment by zooba** [comment]
You can make legitimate filenames that include those characters, unfortunately. They have to be escaped, but they can be valid.

The kind of escaping necessary also depends on what quoting is in the command that the value is being substituted into, so trying to figure that out ahead of time seems to be near impossible.

**Comment by encukou** [comment]
Yes. It's a trade-off between security and functionality.

**Comment by vstinner** [comment]
I'm scared by the PR #91993 approach: either use a blocklist (reject a filename which looks like a shell command) or an allowlist (strict format for filenames).

A blocklist can reject legit filenames which worked previously. For example, Linux accepts semicolon ``;`` in directory and file names. This is perfectly valid:

```
$ echo "hello" > 'a.txt;id'
$ cat 'a.txt;id'
hello
```

A mailcap file can quote properly filenames and so accept such unusual filename.

An allowlist can also reject legit filenames which worked previously. For example, if we reject non-ASCII characters, a non-ASCII filename which worked previously will no longer work:

```
$ echo "hello" > hervé.txt
$ cat hervé.txt
hello
```

Linux filesystems (like ext4 or XFS) obviously accepts non-ASCII filenames.

---

Adding quotes is dangerous:

* Depending on which shell implementation is used to run the command, the executed command may be different (maybe I'm wrong and POSIX is super strict about that?).
* The mailcap file can already contain quotes.
* It's hard to decide how to escape existing quotes in the input filename... Again, quotes are valid characters in a filename...
* There are different kinds of quotes: ``'single'`` and ``"double"``. Double quotes can also embed shell commands: 

```
$ echo "filename$(pwd).txt"
filename/home/vstinner/python/main.txt
```

It's a dangerous whale-a-mole game :-(

---

For all these reasons, I suggest to **not fix the vulnerability**, but **document** it and explains how to **work around** the issue. I proposed the PR #92024 for that.

**Comment by vstinner** [comment]
About mailcap files. Fedora 35 provides:

```
$ cat /etc/mailcap 
### 
### Begin Red Hat Mailcap
###

audio/*; /usr/bin/xdg-open %s

image/*; /usr/bin/xdg-open %s

application/msword; /usr/bin/xdg-open %s
application/pdf; /usr/bin/xdg-open %s
application/postscript ; /usr/bin/xdg-open %s

text/html; /usr/bin/xdg-open %s ; copiousoutput
```

No filename is put between codes.

The Python test suite uses this file:

```
# Mailcap file for test_mailcap; based on RFC 1524
# Referred to by test_mailcap.py

#
# This is a comment.
#

application/frame; showframe %s; print="cat %s | lp"
application/postscript; ps-to-terminal %s;\
    needsterminal
application/postscript; ps-to-terminal %s; \
    compose=idraw %s
application/x-dvi; xdvi %s
application/x-movie; movieplayer %s; compose=moviemaker %s; \
       description="Movie"; \
       x11-bitmap="/usr/lib/Zmail/bitmaps/movie.xbm"
application/*; echo "This is \"%t\" but \
       is 50 \% Greek to me" \; cat %s; copiousoutput

audio/basic; showaudio %s; compose=audiocompose %s; edit=audiocompose %s;\
description="An audio fragment"
audio/* ; /usr/local/bin/showaudio %t

image/rgb; display %s
#image/gif; display %s
image/x-xwindowdump; display %s

# The continuation char shouldn't \
# make a difference in a comment.

message/external-body; showexternal %s %{access-type} %{name} %{site} \
    %{directory} %{mode} %{server}; needsterminal; composetyped = extcompose %s; \
    description="A reference to data stored in an external location"

text/richtext; shownonascii iso-8859-8 -e richtext -p %s; test=test "`echo \
    %{charset} | tr '[A-Z]' '[a-z]'`"  = iso-8859-8; copiousoutput

video/*; animate %s
video/mpeg; mpeg_play %s
```

No command put the filename ``%s`` between quotes.

**Comment by encukou** [comment]
> I'm scared by the PR https://github.com/python/cpython/pull/91993 approach

I'm confused. Most of the criticism doesn't apply to that PR:

>  For example, Linux accepts semicolon ; in directory and file names.

There is no way to use these safely with mailcap, sadly. That's why I propose to reject them.

>  if we reject non-ASCII characters

I do not propose rejecting non-ASCII characters.

> Adding quotes is dangerous:

I do not propose adding quotes.

> No command put the filename %s between quotes.

Right, Fedora and the Python tests don't do that. But see the discussion above -- you *can* have a mailcap line like that.

**Comment by xiaoge1001** [comment]
For Python3.9 and python3.10,  can we use the patch provided by Bernd dietzel or vstinner? Should security be more important than changes in APIs and other aspects?
We can add a switch to accept changes to fix security vulnerabilities by default. Then, the caller can restore the behavior before repair by modifying configuration or change a variable. If so, caller need to accept this security risk and ensure the security of file names.
[1](https://bugs.python.org/file40897/mailcap%20patch.zip)
[2](https://github.com/python/cpython/pull/91993)

**Comment by gpshead** [comment]
The failure scenario for an application where we simply start claiming no match (return None?) on potentially suspicious filenames seems acceptable. I _doubt_ it would be a big deal to users if any even notice at all. So if you want to go forward with a PR like #91993 my gut feeling is that nobody is going to balk at the change, even in security only release branches.

**Comment by jsbueno** [comment]
We (at my company, but I suppose it can be a widespread problem) are facing a problem with this now that a CVE has been opened pointing to this issue. Any container images with a Python package installed is being flagged as containing a "critical vulnerability". 

That said, I just saw there is a Python 3.9 release scheduled for next week. Maybe one of the proposed fixes could be merged and backported before that?

(I don't think any one person running an updated Python would ever notice if a badly quoted call to `mailcap.findmatch`s would simply stop the program. But them, there is https://xkcd.com/1172/ .)

**Comment by zooba** [comment]
> Any container images with a Python package installed is being flagged as containing a "critical vulnerability".

Our apologies. We hesitated to assign the CVE because we knew this would happen, but it does meet the criteria and so the ID could have been assigned anyway.

Hopefully you have some kind of system to dismiss those alerts. If you're not using `mailcap` (which seems likely), then you are not impacted.

**Comment by phbno1** [comment]
@All respectable developers, can you give a clear solution, either remove the module or fix it, this is a 9.8 high-risk vulnerability that can't be hung all the time? You know, all agencies or organizations that use python are waiting for a solution from the official python community.

**Comment by brettcannon** [comment]
@phbno1 the module has been deprecated in Python 3.11 and will be removed in Python 3.13.

**Comment by phbno1** [comment]
> What about the other versions?

**Comment by phbno1** [comment]
> @phbno1 the module has been deprecated in Python 3.11 and will be removed in Python 3.13.

What about the other versions?

**Comment by zooba** [comment]
If you're not using it (then you're not affected, but...) you can delete the file from your own installs. That's at least as easy as installing a full update from us, and far less impactful to your applications.

This is a really easy issue to determine your exposure. If you're stuck because you have tooling that can't accommodate this kind of situation (I've got to deal with it too, JFYI), complain to the tooling vendor.

**Comment by kirotawa** [comment]
Hi there, I'm curious to know if you folks have a good fix candidate for it. I'm a bit confuse as it has at least two fix for it on going? #91993 and #93014

**Comment by yeroc-sebrof** [comment]
> If you're not using it (then you're not affected, but...) you can delete the file from your own installs. That's at least as easy as installing a full update from us, and far less impactful to your applications.
> 
> This is a really easy issue to determine your exposure. If you're stuck because you have tooling that can't accommodate this kind of situation (I've got to deal with it too, JFYI), complain to the tooling vendor.

Agreed, `rm /usr/local/lib/python3.*/mailcap.py` is the easiest option to ensure your safety on tools where you know you're not actually using this dep and want the vulnerability scanners to stop complaining.

**Comment by iritkatriel** [comment]
There are warnings from a few tests, I believe they are related to this issue.

```
% ./python.exe -We -m test -v test_mailcap
[...]
======================================================================
ERROR: test_findmatch (test.test_mailcap.FindmatchTest.test_findmatch)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/iritkatriel/src/cpython-1/Lib/test/test_mailcap.py", line 223, in test_findmatch
    self._run_cases(cases)
  File "/Users/iritkatriel/src/cpython-1/Lib/test/test_mailcap.py", line 250, in _run_cases
    self.assertEqual(mailcap.findmatch(*c[0], **c[1]), c[2])
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/iritkatriel/src/cpython-1/Lib/mailcap.py", line 193, in findmatch
    command = subst(e[key], MIMEtype, filename, plist)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/iritkatriel/src/cpython-1/Lib/mailcap.py", line 230, in subst
    warnings.warn(msg, UnsafeMailcapInput)
mailcap.UnsafeMailcapInput: Refusing to substitute MIME type 'audio/*' into a shell command.

======================================================================
ERROR: test_subst (test.test_mailcap.HelperFunctionTest.test_subst)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/iritkatriel/src/cpython-1/Lib/test/test_mailcap.py", line 137, in test_subst
    self.assertEqual(mailcap.subst(*tc[0]), tc[1])
                     ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/iritkatriel/src/cpython-1/Lib/mailcap.py", line 230, in subst
    warnings.warn(msg, UnsafeMailcapInput)
mailcap.UnsafeMailcapInput: Refusing to substitute MIME type 'audio/*' into a shell command.

----------------------------------------------------------------------

```

**Comment by rajivbandi** [comment]
Will this change be backported to 3.9.x as a security fix ? When is this change expected to be released ?

**Comment by vstinner** [comment]
> Will this change be backported to 3.9.x as a security fix ? When is this change expected to be released ?

Oh right, only 3.11 and main branches are fixed yet. The fix should be backported to other Python versions: https://devguide.python.org/versions/

**Comment by ambv** [comment]
Fixed in 3.7 - 3.12.

**Comment by vstinner** [comment]
I created https://python-security.readthedocs.io/vuln/mailcap-shell-injection.html to track this vulnerability.

**Comment by c3ivodujmovic** [comment]
@ambv can you list the versions in which this is fixed for 3.7, 3.8, 3.9?  The doc by @vstinner still  is missing this info. 
PS the scanner tools are all pointing to this CVE as open.

**Comment by c3ivodujmovic** [comment]
First thank you for your help here and the work on the project.

We also found merge [https://github.com/python/cpython/pull/98192](https://github.com/python/cpython/pull/98192) from Oct 11 plus [above mentioned](https://github.com/AnacondaRecipes/python-feedstock/pull/75) from Oct 19. 

Is there somewhere an explanation of the release flow from merges into [AnacondaRecipes/python-feedstock](https://github.com/AnacondaRecipes/python-feedstock) and [python/cpython](https://github.com/python/cpython) -- all the way to released python versions?  It would make these investigations self-serve. :)   Thank you

**Comment by ambv** [comment]
No idea what Anaconda does, you need to ask them. As for CPython, our process is that [a fix lands in `main` first and then gets progressively backported](https://devguide.python.org/core-developers/committing/#branch-merge). In the case of the fix in question, 3.9 and older backports happened after the [previous round of releases](https://discuss.python.org/t/python-versions-3-10-8-3-9-15-3-8-15-3-7-15-now-available/19889/5) was cut.

Per [PEP 619](https://peps.python.org/pep-0619/) the next 3.10 bugfix release is scheduled for December 5th and the other release managers synchronized their calendars to release 3.7 - 3.12 on that day.

**Comment by ambv** [comment]
And by the way, on this issue you see the PRs for the backports mentioned with their respective branch in links like these:

<img width="882" alt="Screen Shot 2022-11-23 at 19 47 16" src="https://user-images.githubusercontent.com/55281/203624829-76d1549f-39f9-4d3b-9d27-54af233d527b.png">

**Comment by vstinner** [comment]
> @ambv can you list the versions in which this is fixed for 3.7, 3.8, 3.9? The doc by @vstinner still is missing this info.

https://python-security.readthedocs.io/vuln/mailcap-shell-injection.html seems to be up to date, no?

Fixed In

* Python 3.10.8 (2022-10-11) fixed by commit 96739bc (branch 3.10) (2022-09-20)
* Python 3.11.0 (2022-10-24) fixed by commit fae93ab (branch 3.11) (2022-06-03)

Vulnerable Versions

* Python 3.7 (need release)
* Python 3.8 (need release)
* Python 3.9 (need release)

@ambv announced the next batch of releases. In general, I look at release PEPs in https://devguide.python.org/versions/ for estimated release dates.

**Comment by katrielkap** [comment]
Hey, can you update the Known Affected Software Configurations (CPE) in the CVE to the correct one as mentioned below and also mention the older unfixed versions in the CPE as well ?

> Fixed In
> Python 3.10.8 (2022-10-11) fixed by commit 96739bc (branch 3.10) (2022-09-20)
> Python 3.11.0 (2022-10-24) fixed by commit fae93ab (branch 3.11) (2022-06-03)

Also @vstinner , It seems like https://python-security.readthedocs.io/vuln/mailcap-shell-injection.html is not up-to-date - "In Python (aka CPython) through 3.10.4".

**Comment by vstinner** [comment]
> Also @vstinner , It seems like https://python-security.readthedocs.io/vuln/mailcap-shell-injection.html is not up-to-date - "In Python (aka CPython) through 3.10.4".

My tool just copies what the CVE says.
